COOP: test COOP popup from a CSP-sandboxed popup

zcorpan opened the pull request:
Part of #18354.

zcorpan committed:
COOP: test COOP popup from a CSP-sandboxed popup

Part of #18354.

annevk commented on html/cross-origin-opener-policy/resources/csp-sandbox.py:
```diff
@@ -0,0 +1,21 @@
+<script>
+  const params = new URL(location).searchParams;
+  params.delete("sandbox");
+  window.open(`${get_host_info().HTTPS_ORIGIN}/html/cross-origin-opener-policy/resources/coop-coep.py?${params}`)
```
Is this meant to be the same origin?

zcorpan replied:
Hmm, yes, but I realize now that `get_host_info` isn't needed, it can just be a relative URL instead.

annevk reviewed (commented):
It'd be interesting to also navigate the CSP popup to a COOP document as that should work (since it's the document that's sandboxed and not the browsing context).

zcorpan committed:
Apply suggestions from code review

Co-Authored-By: Anne van Kesteren <annevk@annevk.nl>

zcorpan committed:
Remove get-host-info.sub.js from csp-sandbox.py

ParisMeuleman commented:
> It'd be interesting to also navigate the CSP popup to a COOP document as that should work (since it's the document that's sandboxed and not the browsing context).

Thanks for that comment, it helped clarify a doubt in Chromium's implem: https://crrev.com/c/1995268/3/content/browser/cross_origin_opener_policy_browsertest.cc

Hence, I think adding this test is a good idea!

cc: @camillelamy @ArthurSonzogni  @hemeryar

zcorpan committed:
Use byte strings in python

zcorpan committed:
Add a test per annevk's suggestion to navigate from CSP sandbox to COOP doc

zcorpan commented:
@annevk

> It'd be interesting to also navigate the CSP popup to a COOP document as that should work (since it's the document that's sandboxed and not the browsing context).

Now added in 1fc0b75

annevk commented on html/cross-origin-opener-policy/coop-csp-sandbox-navigate.https.html:
```diff
@@ -0,0 +1,25 @@
+      t.step_timeout(() => {
+        assert_unreached('Navigation from CSP sandbox to COOP document failed')
+      }, 1500);
+    }));
```
Since you have a reference to the popup and it won't be replaced, can't you assert something about the popup? That it's not closed and that it's document is no longer same-origin? I guess that doesn't work for the case where allow-same-origin isn't set.

zcorpan replied:
But `closed` should be true after the navigation? The CSP sandbox doesn't have COOP, and then when navigating to the COOP document the browsing context is swapped.

The "is no longer same origin" is interesting. Should it not be same-origin for any of these 2 cases?

annevk replied:
I misread the test.

annevk commented on html/cross-origin-opener-policy/resources/common.js:
```diff
@@ -92,4 +92,16 @@ function run_coop_test_iframe (documentTitle, iframe_origin, popup_origin, popup
+      });
+    });
+  });
+});
```
It seems somewhat bad to start running this everywhere if we only use it in two places.

zcorpan replied:
Fixed in 8db3e7b

zcorpan committed:
Change fullyLoaded to a function

zcorpan committed:
Add more assertions in the navigate test (unsure about correctness)

annevk commented:
This looks good at a high-level, but seems like it still needs conflicts resolved.

zcorpan commented:
I added a same-origin check, and have different results between Firefox and Chrome:

Test | Firefox | Chrome
---|---|---
CSP:  sandbox allow-popups allow-scripts allow-same-origin; CSP sandbox popup  navigate to Cross-Origin-Opener-Policy document should work | Fail | Fail
CSP: sandbox allow-popups allow-scripts; CSP sandbox popup navigate to Cross-Origin-Opener-Policy document should work | Pass | Fail

I'm not sure what is right. Should the popup after navigation (which causes a BC swap) not be same-origin with the opener?

zcorpan committed:
Avoid spaces in window.open second argument

annevk commented:
Looking at `html/cross-origin-opener-policy/coop-csp-sandbox-navigate.https.html` I think Firefox is correct. In particular `assert_throws_dom("SecurityError", () => { popup.document; }, 'same-origin check');` should not throw when the CSP popup has allow-same-origin. Note that popup cannot have a reference to the document being navigated to as it's already closed (and replaced by a new thing you cannot have a reference to by design) at that point.

zcorpan committed:
Correct the same-origin check and add explanatory comments

zcorpan commented:
@annevk thank you for explaining that, I had misremembered how Window/WindowProxy works on navigation. `popup` is the old window which still has the CSP sandbox applied. I've fixed the test and added comments.

annevk commented:
Will you file a bug against Chrome? It seems they made a mistake somewhere (or you/I analyzed incorrectly).

zcorpan commented:
Of course: https://bugs.chromium.org/p/chromium/issues/detail?id=1122582

zcorpan commented:
I'll admin-merge this, see https://github.com/web-platform-tests/wpt/pull/21111#issuecomment-678262168
